File: desktop/src/types/agent.ts
/**
 * Agent Types
 *
 * TypeScript interfaces matching the Rust models in desktop/src-tauri/src/models/agent.rs
 */

/** AI Agent with custom system prompt and tool configuration */
export interface Agent {
  /** Unique identifier (UUID) */
  id: string;
  /** Display name for the agent */
  name: string;
  /** Description of what the agent does */
  description: string | null;
  /** System prompt that defines the agent's behavior */
  system_prompt: string;
  /** Model to use (e.g., "claude-sonnet-4-20250514") */
  model: string;
  /** List of allowed tool names (empty means all tools allowed) */
  allowed_tools: string[];
  /** Creation timestamp (ISO 8601) */
  created_at: string | null;
  /** Last update timestamp (ISO 8601) */
  updated_at: string | null;
}

/** Request to create a new agent */
export interface AgentCreateRequest {
  /** Display name for the agent */
  name: string;
  /** Description of what the agent does */
  description?: string | null;
  /** System prompt that defines the agent's behavior */
  system_prompt: string;
  /** Model to use */
  model: string;
  /** List of allowed tool names */
  allowed_tools: string[];
}

/** Request to update an existing agent */
export interface AgentUpdateRequest {
  /** New display name (optional) */
  name?: string | null;
  /** New description (optional) */
  description?: string | null;
  /** New system prompt (optional) */
  system_prompt?: string | null;
  /** New model (optional) */
  model?: string | null;
  /** New allowed tools list (optional) */
  allowed_tools?: string[] | null;
}

/** Agent run status */
export type RunStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';

/** A single execution run of an agent */
export interface AgentRun {
  /** Unique run identifier */
  id: string;
  /** Agent that was executed */
  agent_id: string;
  /** User input that triggered the run */
  input: string;
  /** Output produced by the agent */
  output: string | null;
  /** Current status of the run */
  status: RunStatus;
  /** Execution duration in milliseconds */
  duration_ms: number | null;
  /** Number of input tokens used */
  input_tokens: number | null;
  /** Number of output tokens generated */
  output_tokens: number | null;
  /** Error message if the run failed */
  error: string | null;
  /** When the run was created */
  created_at: string | null;
  /** When the run completed */
  completed_at: string | null;
}

/** Statistics for an agent */
export interface AgentStats {
  /** Total number of runs */
  total_runs: number;
  /** Number of successful runs */
  completed_runs: number;
  /** Number of failed runs */
  failed_runs: number;
  /** Number of cancelled runs */
  cancelled_runs: number;
  /** Success rate (completed / total) as percentage */
  success_rate: number;
  /** Average execution duration in milliseconds */
  avg_duration_ms: number;
  /** Total input tokens used */
  total_input_tokens: number;
  /** Total output tokens generated */
  total_output_tokens: number;
  /** Timestamp of the last run */
  last_run_at: string | null;
}

/** Paginated list of agent runs */
export interface AgentRunList {
  /** List of runs */
  runs: AgentRun[];
  /** Total count of runs (for pagination) */
  total: number;
  /** Current offset */
  offset: number;
  /** Number of items per page */
  limit: number;
}

/** Agent with statistics (for UI display) */
export interface AgentWithStats extends Agent {
  stats: AgentStats;
}

/** Generic command response from Tauri */
export interface CommandResponse<T> {
  success: boolean;
  data: T | null;
  error: string | null;
}

/** Available Claude models */
export const CLAUDE_MODELS = [
  { id: 'claude-sonnet-4-20250514', name: 'Claude Sonnet 4', description: 'Balanced performance' },
  { id: 'claude-opus-4-20250514', name: 'Claude Opus 4', description: 'Most capable' },
  { id: 'claude-3-5-sonnet-20241022', name: 'Claude 3.5 Sonnet', description: 'Previous generation' },
  { id: 'claude-3-5-haiku-20241022', name: 'Claude 3.5 Haiku', description: 'Fast and efficient' },
] as const;

/** Default model for new agents */
export const DEFAULT_MODEL = 'claude-sonnet-4-20250514';

/** Available tools that can be assigned to agents */
export const AVAILABLE_TOOLS = [
  // File operations
  { name: 'read_file', category: 'File Operations', description: 'Read file contents' },
  { name: 'write_file', category: 'File Operations', description: 'Write to a file' },
  { name: 'edit_file', category: 'File Operations', description: 'Edit file contents' },
  { name: 'list_directory', category: 'File Operations', description: 'List directory contents' },
  { name: 'search_files', category: 'File Operations', description: 'Search for files' },
  // Code operations
  { name: 'execute_code', category: 'Code', description: 'Execute code snippets' },
  { name: 'run_tests', category: 'Code', description: 'Run test suites' },
  { name: 'lint_code', category: 'Code', description: 'Lint and format code' },
  // Git operations
  { name: 'git_status', category: 'Git', description: 'Get git status' },
  { name: 'git_diff', category: 'Git', description: 'Show git diff' },
  { name: 'git_commit', category: 'Git', description: 'Create git commits' },
  { name: 'git_push', category: 'Git', description: 'Push to remote' },
  // Web operations
  { name: 'web_search', category: 'Web', description: 'Search the web' },
  { name: 'web_fetch', category: 'Web', description: 'Fetch web content' },
  // System operations
  { name: 'execute_command', category: 'System', description: 'Execute shell commands' },
] as const;

/** Get tools grouped by category */
export function getToolsByCategory(): Record<string, typeof AVAILABLE_TOOLS[number][]> {
  const grouped: Record<string, typeof AVAILABLE_TOOLS[number][]> = {};

  for (const tool of AVAILABLE_TOOLS) {
    if (!grouped[tool.category]) {
      grouped[tool.category] = [];
    }
    grouped[tool.category].push(tool);
  }

  return grouped;
}

/** Format duration in a human-readable way */
export function formatDuration(ms: number | null): string {
  if (ms === null || ms === undefined) return '-';

  if (ms < 1000) {
    return `${ms}ms`;
  } else if (ms < 60000) {
    return `${(ms / 1000).toFixed(1)}s`;
  } else {
    const minutes = Math.floor(ms / 60000);
    const seconds = ((ms % 60000) / 1000).toFixed(0);
    return `${minutes}m ${seconds}s`;
  }
}

/** Format token count in a human-readable way */
export function formatTokens(count: number | null): string {
  if (count === null || count === undefined) return '-';

  if (count < 1000) {
    return count.toString();
  } else if (count < 1000000) {
    return `${(count / 1000).toFixed(1)}K`;
  } else {
    return `${(count / 1000000).toFixed(2)}M`;
  }
}

/** Get status color for UI display */
export function getStatusColor(status: RunStatus): string {
  switch (status) {
    case 'completed':
      return 'text-green-600 dark:text-green-400';
    case 'running':
      return 'text-blue-600 dark:text-blue-400';
    case 'pending':
      return 'text-yellow-600 dark:text-yellow-400';
    case 'failed':
      return 'text-red-600 dark:text-red-400';
    case 'cancelled':
      return 'text-gray-600 dark:text-gray-400';
    default:
      return 'text-gray-500';
  }
}

/** Get status badge background color */
export function getStatusBgColor(status: RunStatus): string {
  switch (status) {
    case 'completed':
      return 'bg-green-100 dark:bg-green-900/50';
    case 'running':
      return 'bg-blue-100 dark:bg-blue-900/50';
    case 'pending':
      return 'bg-yellow-100 dark:bg-yellow-900/50';
    case 'failed':
      return 'bg-red-100 dark:bg-red-900/50';
    case 'cancelled':
      return 'bg-gray-100 dark:bg-gray-900/50';
    default:
      return 'bg-gray-100 dark:bg-gray-800';
  }
}

File: desktop/src-tauri/src/models/agent.rs
//! Agent Models
//!
//! Data structures for AI agents with custom behaviors.

use serde::{Deserialize, Serialize};

/// An AI agent with custom system prompt and tool configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Agent {
    /// Unique identifier (UUID)
    pub id: String,
    /// Display name for the agent
    pub name: String,
    /// Description of what the agent does
    pub description: Option<String>,
    /// System prompt that defines the agent's behavior
    pub system_prompt: String,
    /// Model to use (e.g., "claude-sonnet-4-20250514")
    pub model: String,
    /// List of allowed tool names (empty means all tools allowed)
    pub allowed_tools: Vec<String>,
    /// Creation timestamp (ISO 8601)
    pub created_at: Option<String>,
    /// Last update timestamp (ISO 8601)
    pub updated_at: Option<String>,
}

impl Agent {
    /// Create a new agent with required fields
    pub fn new(
        id: impl Into<String>,
        name: impl Into<String>,
        system_prompt: impl Into<String>,
        model: impl Into<String>,
    ) -> Self {
        let now = chrono::Utc::now().to_rfc3339();
        Self {
            id: id.into(),
            name: name.into(),
            description: None,
            system_prompt: system_prompt.into(),
            model: model.into(),
            allowed_tools: Vec::new(),
            created_at: Some(now.clone()),
            updated_at: Some(now),
        }
    }

    /// Create a new agent with a generated UUID
    pub fn create(
        name: impl Into<String>,
        system_prompt: impl Into<String>,
        model: impl Into<String>,
    ) -> Self {
        Self::new(uuid::Uuid::new_v4().to_string(), name, system_prompt, model)
    }

    /// Builder pattern: set description
    pub fn with_description(mut self, description: impl Into<String>) -> Self {
        self.description = Some(description.into());
        self
    }

    /// Builder pattern: set allowed tools
    pub fn with_allowed_tools(mut self, tools: Vec<String>) -> Self {
        self.allowed_tools = tools;
        self
    }
}

/// Request to create a new agent
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentCreateRequest {
    /// Display name for the agent
    pub name: String,
    /// Description of what the agent does
    pub description: Option<String>,
    /// System prompt that defines the agent's behavior
    pub system_prompt: String,
    /// Model to use
    pub model: String,
    /// List of allowed tool names
    pub allowed_tools: Vec<String>,
}

impl AgentCreateRequest {
    /// Validate the create request
    pub fn validate(&self) -> Result<(), String> {
        if self.name.trim().is_empty() {
            return Err("Agent name cannot be empty".to_string());
        }
        if self.name.len() > 100 {
            return Err("Agent name cannot exceed 100 characters".to_string());
        }
        if self.system_prompt.trim().is_empty() {
            return Err("System prompt cannot be empty".to_string());
        }
        if self.system_prompt.len() > 100_000 {
            return Err("System prompt cannot exceed 100,000 characters".to_string());
        }
        if self.model.trim().is_empty() {
            return Err("Model cannot be empty".to_string());
        }
        Ok(())
    }
}

/// Request to update an existing agent
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentUpdateRequest {
    /// New display name (optional)
    pub name: Option<String>,
    /// New description (optional)
    pub description: Option<String>,
    /// New system prompt (optional)
    pub system_prompt: Option<String>,
    /// New model (optional)
    pub model: Option<String>,
    /// New allowed tools list (optional)
    pub allowed_tools: Option<Vec<String>>,
}

impl AgentUpdateRequest {
    /// Validate the update request
    pub fn validate(&self) -> Result<(), String> {
        if let Some(ref name) = self.name {
            if name.trim().is_empty() {
                return Err("Agent name cannot be empty".to_string());
            }
            if name.len() > 100 {
                return Err("Agent name cannot exceed 100 characters".to_string());
            }
        }
        if let Some(ref system_prompt) = self.system_prompt {
            if system_prompt.trim().is_empty() {
                return Err("System prompt cannot be empty".to_string());
            }
            if system_prompt.len() > 100_000 {
                return Err("System prompt cannot exceed 100,000 characters".to_string());
            }
        }
        if let Some(ref model) = self.model {
            if model.trim().is_empty() {
                return Err("Model cannot be empty".to_string());
            }
        }
        Ok(())
    }

    /// Check if the request has any fields to update
    pub fn has_updates(&self) -> bool {
        self.name.is_some()
            || self.description.is_some()
            || self.system_prompt.is_some()
            || self.model.is_some()
            || self.allowed_tools.is_some()
    }
}

/// Agent run status
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum RunStatus {
    /// Run is pending execution
    Pending,
    /// Run is currently executing
    Running,
    /// Run completed successfully
    Completed,
    /// Run failed with an error
    Failed,
    /// Run was cancelled by user
    Cancelled,
}

impl std::fmt::Display for RunStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Pending => write!(f, "pending"),
            Self::Running => write!(f, "running"),
            Self::Completed => write!(f, "completed"),
            Self::Failed => write!(f, "failed"),
            Self::Cancelled => write!(f, "cancelled"),
        }
    }
}

impl std::str::FromStr for RunStatus {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "pending" => Ok(Self::Pending),
            "running" => Ok(Self::Running),
            "completed" => Ok(Self::Completed),
            "failed" => Ok(Self::Failed),
            "cancelled" => Ok(Self::Cancelled),
            _ => Err(format!("Unknown run status: {}", s)),
        }
    }
}

/// A single execution run of an agent
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentRun {
    /// Unique run identifier
    pub id: String,
    /// Agent that was executed
    pub agent_id: String,
    /// User input that triggered the run
    pub input: String,
    /// Output produced by the agent
    pub output: Option<String>,
    /// Current status of the run
    pub status: RunStatus,
    /// Execution duration in milliseconds
    pub duration_ms: Option<u64>,
    /// Number of input tokens used
    pub input_tokens: Option<u32>,
    /// Number of output tokens generated
    pub output_tokens: Option<u32>,
    /// Error message if the run failed
    pub error: Option<String>,
    /// When the run was created
    pub created_at: Option<String>,
    /// When the run completed
    pub completed_at: Option<String>,
}

impl AgentRun {
    /// Create a new pending run
    pub fn new(agent_id: impl Into<String>, input: impl Into<String>) -> Self {
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            agent_id: agent_id.into(),
            input: input.into(),
            output: None,
            status: RunStatus::Pending,
            duration_ms: None,
            input_tokens: None,
            output_tokens: None,
            error: None,
            created_at: Some(chrono::Utc::now().to_rfc3339()),
            completed_at: None,
        }
    }

    /// Mark the run as started
    pub fn start(&mut self) {
        self.status = RunStatus::Running;
    }

    /// Mark the run as completed with output
    pub fn complete(&mut self, output: String, duration_ms: u64, input_tokens: u32, output_tokens: u32) {
        self.status = RunStatus::Completed;
        self.output = Some(output);
        self.duration_ms = Some(duration_ms);
        self.input_tokens = Some(input_tokens);
        self.output_tokens = Some(output_tokens);
        self.completed_at = Some(chrono::Utc::now().to_rfc3339());
    }

    /// Mark the run as failed with an error
    pub fn fail(&mut self, error: String, duration_ms: u64) {
        self.status = RunStatus::Failed;
        self.error = Some(error);
        self.duration_ms = Some(duration_ms);
        self.completed_at = Some(chrono::Utc::now().to_rfc3339());
    }

    /// Mark the run as cancelled
    pub fn cancel(&mut self, duration_ms: u64) {
        self.status = RunStatus::Cancelled;
        self.duration_ms = Some(duration_ms);
        self.completed_at = Some(chrono::Utc::now().to_rfc3339());
    }
}

/// Statistics for an agent
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct AgentStats {
    /// Total number of runs
    pub total_runs: u32,
    /// Number of successful runs
    pub completed_runs: u32,
    /// Number of failed runs
    pub failed_runs: u32,
    /// Number of cancelled runs
    pub cancelled_runs: u32,
    /// Success rate (completed / total) as percentage
    pub success_rate: f64,
    /// Average execution duration in milliseconds
    pub avg_duration_ms: f64,
    /// Total input tokens used
    pub total_input_tokens: u64,
    /// Total output tokens generated
    pub total_output_tokens: u64,
    /// Timestamp of the last run
    pub last_run_at: Option<String>,
}

impl AgentStats {
    /// Calculate success rate
    pub fn calculate_success_rate(&mut self) {
        if self.total_runs > 0 {
            self.success_rate = (self.completed_runs as f64 / self.total_runs as f64) * 100.0;
        } else {
            self.success_rate = 0.0;
        }
    }
}

/// Paginated list of agent runs
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentRunList {
    /// List of runs
    pub runs: Vec<AgentRun>,
    /// Total count of runs (for pagination)
    pub total: u32,
    /// Current offset
    pub offset: u32,
    /// Number of items per page
    pub limit: u32,
}

/// Agent with statistics (for UI display)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentWithStats {
    #[serde(flatten)]
    pub agent: Agent,
    pub stats: AgentStats,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_agent_creation() {
        let agent = Agent::create("Test Agent", "You are a helpful assistant.", "claude-sonnet-4-20250514");
        assert!(!agent.id.is_empty());
        assert_eq!(agent.name, "Test Agent");
        assert_eq!(agent.system_prompt, "You are a helpful assistant.");
        assert!(agent.created_at.is_some());
    }

    #[test]
    fn test_agent_builder() {
        let agent = Agent::create("Test", "Prompt", "model")
            .with_description("A test agent")
            .with_allowed_tools(vec!["read_file".to_string(), "write_file".to_string()]);

        assert_eq!(agent.description, Some("A test agent".to_string()));
        assert_eq!(agent.allowed_tools.len(), 2);
    }

    #[test]
    fn test_create_request_validation() {
        let valid = AgentCreateRequest {
            name: "Test".to_string(),
            description: None,
            system_prompt: "Prompt".to_string(),
            model: "model".to_string(),
            allowed_tools: vec![],
        };
        assert!(valid.validate().is_ok());

        let empty_name = AgentCreateRequest {
            name: "".to_string(),
            description: None,
            system_prompt: "Prompt".to_string(),
            model: "model".to_string(),
            allowed_tools: vec![],
        };
        assert!(empty_name.validate().is_err());
    }

    #[test]
    fn test_run_status_parsing() {
        assert_eq!("pending".parse::<RunStatus>().unwrap(), RunStatus::Pending);
        assert_eq!("COMPLETED".parse::<RunStatus>().unwrap(), RunStatus::Completed);
        assert!("invalid".parse::<RunStatus>().is_err());
    }

    #[test]
    fn test_agent_run_lifecycle() {
        let mut run = AgentRun::new("agent-1", "Hello");
        assert_eq!(run.status, RunStatus::Pending);

        run.start();
        assert_eq!(run.status, RunStatus::Running);

        run.complete("World".to_string(), 1000, 10, 5);
        assert_eq!(run.status, RunStatus::Completed);
        assert_eq!(run.output, Some("World".to_string()));
        assert!(run.completed_at.is_some());
    }

    #[test]
    fn test_agent_stats() {
        let mut stats = AgentStats {
            total_runs: 10,
            completed_runs: 8,
            failed_runs: 2,
            ..Default::default()
        };
        stats.calculate_success_rate();
        assert!((stats.success_rate - 80.0).abs() < 0.01);
    }
}
